### PR TITLE
Added non-trivial move operations.

### PR DESCRIPTION
**Pull request by Jim-Marsden**
* [x] Does the code follow the SFML [Code Style Guide](https://www.sfml-dev.org/style.php)?
* [x] Have you provided some example/test code for your changes?

----

## Description
As noted below there was an issue with my Shader PR. I created explicit move constructors/assignment operators.

This PR is related to the issue #
[https://github.com/SFML/SFML/pull/2273#issuecomment-1321167992](https://github.com/SFML/SFML/pull/2273#issuecomment-1321167992)

## Tasks

* [ ] Tested on Linux
* [ ] Tested on Windows
* [x] Tested on macOS
* [ ] Tested on iOS
* [ ] Tested on Android

## How to test this PR?

Describe how to best test these changes. Please provide a [minimal, complete and verifiable example](https://stackoverflow.com/help/mcve) if possible, you can use the follow template as a start:

```cpp
#include <fmt/format.h>
#include <SFML/Graphics.hpp>

constexpr auto texture_loc{"data/textures/test.png"};
constexpr auto shader_loc{"data/shaders/v.vert"};

auto make_sprite(){
    struct {sf::Texture texture; sf::Sprite sprite;} result;

    if(not result.texture.loadFromFile(texture_loc)){
        throw std::runtime_error(fmt::format("Unable to open file: {}\n", texture_loc));
    }

    result.sprite.setTexture(result.texture);
    return result;
}

auto make_shader(){
    sf::Shader shader;
    if(not shader.loadFromFile(shader_loc, sf::Shader::Vertex)){
        throw std::runtime_error(fmt::format("Unable to open shader: {}\n", shader_loc));
    }
    shader.setUniform("target_color", sf::Vector3f{0, 1, 1});
    return shader;
}

int main() try {
    sf::RenderWindow window{sf::VideoMode{{128, 128}}, "shader move test"};

     // to cut down on boilerplate
    auto [texture, sprite] = make_sprite(); 
    auto shader{make_shader()};

    while(window.isOpen()){
        for(sf::Event event{}; window.pollEvent(event);){
            if(sf::Event::Closed == event.type){
                window.close();
            }
        }

        // Currently both shaders have the same shader native handle after the move.
        sf::Shader move_test = std::move(shader);
        if(move_test.getNativeHandle() == shader.getNativeHandle())
        {
            // Throws on master, fixed with this PR. It tries to free the shader object twice.
            throw std::runtime_error("The shaders share a native handle.");
        }
        window.clear();
        window.draw(sprite, {&move_test});
        window.display();

         // So we can move again on the next loop.
        shader = std::move(move_test);
    } // If we don't throw we delete the shader object and on the next cycle try to use an invalid shader.
}
catch (std::runtime_error const & e){
    fmt::print(stderr, "Exception caught: {}\n", e.what());
}
```

### CODE REVIEW

**Comment by eXpl0it3r** [comment]
SFMLBuildMaster: Build this please.

**Comment by Jim-Marsden** [comment]
I am more than happy to to wait and see about [https://github.com/SFML/SFML/pull/2280[(https://github.com/SFML/SFML/pull/2280). I think it may be a better solution overall.

**Comment by Bromeon** [comment]
Labeled as `s:undecided` until we have more clarity about #2280.

**Comment by Bambo-Borris** [comment]
Given #2280 is still a draft and there is a known issue with the previous move PR for `sf::Shader` (as identified by @kimci86 on #2273) is it worth getting this in to resolve the problem and then we see later on how the draft PR from @ChrisThrasher progresses?

**Comment by JonnyPtn** [comment]
Yeah I agree - unique resource could take some time before it's suitable (if it ever is) and this is a good change regardless

**Comment by eXpl0it3r** [comment]
I agree, let's continue with this for now, so we don't leave things in a broken state, regardless of whether we find a different solution later on.

**Review by ChrisThrasher — APPROVED** [review]
Thank you for fixing this!

File: include/SFML/Graphics/Shader.hpp [hunk 1]
```diff
@@ -112,16 +112,16 @@ class SFML_GRAPHICS_API Shader : GlResource
     Shader& operator=(const Shader&) = delete;
 
     ////////////////////////////////////////////////////////////
-    /// \brief Defaulted move constructor
+    /// \brief move constructor
```
Comment (eXpl0it3r) [inline include/SFML/Graphics/Shader.hpp:115]: **M**ove...

File: include/SFML/Graphics/Shader.hpp [hunk 2]
```diff
@@ -112,16 +112,16 @@ class SFML_GRAPHICS_API Shader : GlResource
...
     ///
     ////////////////////////////////////////////////////////////
-    Shader(Shader&&) noexcept = default;
+    Shader(Shader&& source) noexcept;
 
     ////////////////////////////////////////////////////////////
     /// \brief Defaulted move assignment
```
Comment (eXpl0it3r) [inline include/SFML/Graphics/Shader.hpp:121]: ~~Defaulted~~ **M**ove...

**Review by eXpl0it3r — CHANGES_REQUESTED** [review]
Small comment changes

File: src/SFML/Graphics/Shader.cpp [hunk 3]
```diff
@@ -250,6 +250,38 @@ Shader::~Shader()
         glCheck(GLEXT_glDeleteObject(castToGlHandle(m_shaderProgram)));
 }
 
+////////////////////////////////////////////////////////////
+Shader::Shader(Shader&& source) noexcept :
+m_shaderProgram(std::exchange(source.m_shaderProgram, 0U)),
+m_currentTexture(std::exchange(source.m_currentTexture, 0)),
```
Comment (kimci86) [inline src/SFML/Graphics/Shader.cpp:256]: -1 is the default value so it probably makes sense to assign -1 to the moved-from texture index.
```suggestion
m_currentTexture(std::exchange(source.m_currentTexture, -1)),
```

File: src/SFML/Graphics/Shader.cpp [hunk 4]
```diff
@@ -250,6 +250,38 @@ Shader::~Shader()
...
+    {
+        // Destroy effect program
+        TransientContextLock lock;
+        if (m_shaderProgram)
+            glCheck(GLEXT_glDeleteObject(castToGlHandle(m_shaderProgram)));
+    }
+
+    // Move the contains of right.
+    m_shaderProgram  = std::exchange(right.m_shaderProgram, 0U);
+    m_currentTexture = std::exchange(right.m_currentTexture, 0);
```
Comment (kimci86) [inline src/SFML/Graphics/Shader.cpp:280]: Same,
```suggestion
    m_currentTexture = std::exchange(right.m_currentTexture, -1);
```

File: src/SFML/Graphics/Shader.cpp [hunk 5]
```diff
@@ -250,6 +250,38 @@ Shader::~Shader()
...
+Shader::Shader(Shader&& source) noexcept :
+m_shaderProgram(std::exchange(source.m_shaderProgram, 0U)),
+m_currentTexture(std::exchange(source.m_currentTexture, 0)),
+m_textures(std::move(source.m_textures)),
+m_uniforms(std::move(source.m_uniforms))
+{
+}
+
+////////////////////////////////////////////////////////////
+Shader& Shader::operator=(sf::Shader&& right) noexcept
```
Comment (kimci86) [inline src/SFML/Graphics/Shader.cpp:263]: Nitpicking :wink:
```suggestion
Shader& Shader::operator=(Shader&& right) noexcept
```

Comment (Jim-Marsden) [reply src/SFML/Graphics/Shader.cpp:263]: Fair enough though!

File: src/SFML/Graphics/Shader.cpp (hunk 3 above)
Comment (Jim-Marsden) [reply src/SFML/Graphics/Shader.cpp]: Good catch, my bad!

File: src/SFML/Graphics/Shader.cpp (hunk 4 above)
Comment (Jim-Marsden) [reply src/SFML/Graphics/Shader.cpp]: Same!

File: src/SFML/Graphics/Shader.cpp [hunk 6]
```diff
@@ -250,6 +250,38 @@ Shader::~Shader()
...
+    }
+    // Explicit scope for RAII
+    {
+        // Destroy effect program
+        TransientContextLock lock;
+        if (m_shaderProgram)
+            glCheck(GLEXT_glDeleteObject(castToGlHandle(m_shaderProgram)));
+    }
+
+    // Move the contains of right.
```
Comment (kimci86) [inline src/SFML/Graphics/Shader.cpp:278]: Don't you mean "contents" instead of "contains"?

Comment (Jim-Marsden) [reply src/SFML/Graphics/Shader.cpp:278]: This is awkward, yes.

**Comment by eXpl0it3r** [comment]
SFMLBuildMaster: Build this please.

**Review by eXpl0it3r — APPROVED** [review]

**Pull request merged**